runtime/service/src/main/java/org/apache/polaris/service/admin/PolarisAdminService.java

```diff
@@ -1088,6 +1089,69 @@ public void deletePrincipal(String name) {
     return returnedEntity;
   }
 
+  private @Nonnull PrincipalWithCredentials resetCredentialsHelper(
+      String principalName, String customClientId, String customClientSecret) {
+    PrincipalEntity currentPrincipalEntity =
+        findPrincipalByName(principalName)
+            .orElseThrow(() -> new NotFoundException("Principal %s not found", principalName));
+
+    if (FederatedEntities.isFederated(currentPrincipalEntity)) {
+      throw new ValidationException(
+          "Cannot reset credentials for a federated principal: %s", principalName);
+    }
+    PolarisPrincipalSecrets currentSecrets =
+        metaStoreManager
+            .loadPrincipalSecrets(getCurrentPolarisContext(), currentPrincipalEntity.getClientId())
+            .getPrincipalSecrets();
+    if (currentSecrets == null) {
+      throw new IllegalArgumentException(
+          String.format("Failed to load current secrets for principal '%s'", principalName));
+    }
+    PolarisPrincipalSecrets newSecrets =
+        metaStoreManager
+            .resetPrincipalSecrets(
+                getCurrentPolarisContext(),
+                currentPrincipalEntity.getClientId(),
+                currentPrincipalEntity.getId(),
+                currentSecrets.getMainSecretHash(),
+                customClientId,
+                customClientSecret)
+            .getPrincipalSecrets();
+    if (newSecrets == null) {
+      throw new IllegalStateException(
+          String.format("Failed to %s secrets for principal '%s'", "reset", principalName));
+    }
+    PolarisEntity newPrincipal =
+        PolarisEntity.of(
+            metaStoreManager.loadEntity(
+                getCurrentPolarisContext(),
+                0L,
+                currentPrincipalEntity.getId(),
+                currentPrincipalEntity.getType()));
+
+    PrincipalEntity newPrincipalEntity = PrincipalEntity.of(newPrincipal);
+    if (customClientId != null) {
+      PrincipalEntity.Builder updateBuilder = new PrincipalEntity.Builder(newPrincipalEntity);
+      updateBuilder.setClientId(newSecrets.getPrincipalClientId());
+      PrincipalEntity updatedNewPrincipalEntity = updateBuilder.build();
+      updatedNewPrincipalEntity =
+          Optional.ofNullable(
+                  PrincipalEntity.of(
+                      PolarisEntity.of(
+                          metaStoreManager.updateEntityPropertiesIfNotChanged(
```

Comment:
This is risky WRT `metaStoreManager.resetPrincipalSecrets()` (line 1112). If the request supplies a new client ID and `resetPrincipalSecrets` succeeds, but `updateEntityPropertiesIfNotChanged` fails, then the principal entity's client ID will be out of sync with the `PolarisPrincipalSecrets` client ID. This will likely break all subsequent authentication and reset credentials attempts.

I'd propose to make the inner "reset" methods robust in case the credential is not found by client ID (i.e. create a new credential / hash).

Also, I'd update the entity first as it is more likely to hit conflicts than the credential reset itself.